src/dune_engine/dune_project.ml

```diff
@@ -762,219 +764,236 @@ let parse ~dir ~lang ~opam_packages ~file ~dir_status =
+              ~default:(expand_aliases_in_sandbox_default ~lang)
+          in
+          let root = dir in
+          let file_key = File_key.make ~name ~root in
+          let dialects =
+            List.fold_left
+              ~f:(fun dialects (loc, dialect) ->
+                Dialect.DB.add dialects ~loc dialect)
+              ~init:Dialect.DB.builtin dialects
+          in
+          let () =
+            match name with
+            | Named _ -> ()
+            | Anonymous _ ->
+              if
+                dune_version >= (2, 8)
+                && generate_opam_files
+                && dir_status = Sub_dirs.Status.Normal
+              then
+                let loc = Loc.in_file (Path.source file) in
+                User_warning.emit ~loc
+                  [ Pp.text
+                      "Project name is not specified. Add a (name \
+                       <project-name>) field to your dune-project file to make \
+                       sure that $ dune subst works in release or pinned \
+                       builds"
+                  ]
+          in
+          { name
+          ; file_key
+          ; root
+          ; version
+          ; dune_version
+          ; info
+          ; packages
+          ; stanza_parser
+          ; project_file = file
+          ; extension_args
+          ; parsing_context
+          ; implicit_transitive_deps
+          ; wrapped_executables
+          ; executables_implicit_empty_intf
+          ; accept_alternative_dune_file_name
+          ; generate_opam_files
+          ; use_standard_c_and_cxx_flags
+          ; dialects
+          ; explicit_js_mode
+          ; format_config
+          ; subst_config
+          ; strict_package_deps
+          ; cram
+          ; expand_aliases_in_sandbox
+          }))
+
+let load_dune_project ~dir opam_packages ~dir_status : t Memo.t =
   let file = Path.Source.relative dir filename in
-  load_exn (Path.source file) ~f:(fun lang ->
-      parse ~dir ~lang ~opam_packages ~file ~dir_status)
+  let open Memo.O in
+  let* f =
+    Fs_memo.with_lexbuf_from_file (Path.source file) ~f:(fun lexbuf ->
+        parse_contents lexbuf ~f:(fun lang ->
+            parse ~dir ~lang ~file ~dir_status))
+  in
+  f opam_packages
 
-let load ~dir ~files ~infer_from_opam_files ~dir_status =
+let load ~dir ~files ~infer_from_opam_files ~dir_status : t option Memo.t =
+  let open Memo.O in
   let opam_packages =
     String.Set.fold files ~init:[] ~f:(fun fn acc ->
         match Package.Name.of_opam_file_basename fn with
         | None -> acc
         | Some name ->
           let opam_file = Path.Source.relative dir fn in
           let loc = Loc.in_file (Path.source opam_file) in
-          (* We are lazy here because if the opam files are generated by dune,
-             there is no point parsing them. This value will only be forced if
-             generation of opam files is disabled. *)
-          let pkg = lazy (Package.load_opam_file opam_file name) in
```

Comment:
Why did you remove the laziness? Do you think it wasn't useful?

Reply:
I think the work is still delayed because we aren't binding on the return Memo.t value here. It is left implicit that we are going to be binding on them only once later though. I could add back the lazy if you think it's more clear.

Reply:
Oh, you are right. Sorry.